solvers/wscd.py

```diff
@@ -0,0 +1,183 @@
+        norms[j] = np.sqrt(tmp)
+    return norms
+
+
+@njit
+def cd(X, y, lmbd, gamma, lipschitz, n_iter, winit=None, tol=1e-3):
+    n_samples, n_features = X.shape
+    R = np.copy(y)
+    w = np.zeros(n_features)
+    all_feats = np.arange(n_features)
+    for _ in range(n_iter):
+        for j in range(n_features):
+            if lipschitz[j]:
+                old = w[j]
+                w[j] = prox_mcp(
+                    w[j] + X[:, j] @ R / (lipschitz[j] * n_samples),
+                    lmbd / lipschitz[j],
+                    gamma * lipschitz[j],
+                )
+                diff = old - w[j]
+                if diff != 0:
+                    R += diff * X[:, j]
+        grad = - X.T.dot(y - X.dot(w)) / n_samples
+        subdiff_dist = subdiff_distance(w, grad, all_feats, lmbd, gamma)
+        if np.max(subdiff_dist) < tol:
+            return w
+    return w
+
+
+@njit
+def subdiff_distance(w, grad, ws, lmbd, gamma):
+    """Compute distance of negative gradient to the subdifferential at w."""
+    subdiff_dist = np.zeros_like(grad)
+    for idx, j in enumerate(ws):
+        if w[j] == 0:
+            # distance of -grad to alpha * [-1, 1]
+            subdiff_dist[idx] = max(0, np.abs(grad[idx]) - lmbd)
+        elif np.abs(w[j]) < lmbd * gamma:
+            # distance of -grad_j to (alpha - abs(w[j])/gamma) * sign(w[j])
+            subdiff_dist[idx] = np.abs(
+                grad[idx] + lmbd * np.sign(w[j])
+                - w[j] / gamma)
+        else:
+            # distance of grad to 0
+            subdiff_dist[idx] = np.abs(grad[idx])
+    return subdiff_dist
+
+
+@njit
+def wscd(X, y, lmbd, gamma, lipschitz, n_iter, n_iter_outer, pruning=True,
+         tol=1e-8):
+
+    n_features = X.shape[1]
+    nb_feat_init = 10
+    nb_feat_2_add = 30
+    ind = np.argsort(-np.abs(X.T.dot(y)))[:nb_feat_init]
+
+    all_feats = np.arange(n_features)
+    winit = np.zeros((nb_feat_init))
+    w = np.zeros((n_features))
+
+    for i in range(n_iter_outer):
+        Xaux = X[:, ind]
+
+        w_inter = cd(Xaux, y, lmbd, gamma, lipschitz, n_iter, winit)
+
+        # pruning
+        if pruning:
+            w_inter = w_inter[w_inter.nonzero()[0]]
+            ind = ind[w_inter.nonzero()[0]]
+            Xaux = Xaux[:, w_inter.nonzero()[0]]
+
+        res = y - Xaux @ w_inter
+        grad = - X.T @ res
+
+        subdiff_dist = subdiff_distance(w, grad, all_feats, lmbd, gamma)
+        if np.max(subdiff_dist) < tol:
+            return w
+
+        candidate = np.argsort(-np.abs(grad))
+        nb_add = 0
+        for cand in candidate:
+            if cand not in ind:
+                ind = np.hstack((ind, np.array([cand])))
+                nb_add += 1
+                if nb_add == nb_feat_2_add:
+                    break
+
+        w_init = np.hstack((w_inter, np.zeros(nb_add)))
+
+    w = np.zeros(n_features)
+    w[ind] = w_init
+
+    return w
+
+
+class Solver(BaseSolver):
+    name = "working set cd"
```

Comment:
having names in the solvers name can be problematic when using the command line interface, I suggest to use underscores instead